plugin/fallback/fallback_test.go

```diff
@@ -0,0 +1,226 @@
+type dummyUpstream struct {
+	rcode int
+}
+
+func (u dummyUpstream) From() string {
+	return ""
+}
+
+func (u dummyUpstream) Select() *healthcheck.UpstreamHost {
+	return nil
+}
+
+func (u dummyUpstream) IsAllowedDomain(s string) bool {
+	return false
+}
+
+func (u dummyUpstream) Exchanger() proxy.Exchanger {
+	return dummyExchanger{}
+}
+
+func (u dummyUpstream) Stop() error {
+	return nil
+}
+
+// dummyExchanger implements the proxy.Exchanger interface
+// It is used solely to implement the dummyUpstream above
+type dummyExchanger struct{}
+
+func (e dummyExchanger) Exchange(ctx context.Context, addr string, state request.Request) (*dns.Msg, error) {
+	return &dns.Msg{}, nil
+}
+
+func (e dummyExchanger) Protocol() string {
+	return ""
+}
+
+func (e dummyExchanger) Transport() string {
+	return ""
+}
+
+func (e dummyExchanger) OnStartup(p *proxy.Proxy) error {
+	return nil
+}
+
+func (e dummyExchanger) OnShutdown(p *proxy.Proxy) error {
+	return nil
+}
+
+// Test case for fallback
+type fallbackTestCase struct {
+	rcode            int            // rcode to be returned by the stub Handler
+	expectedUpstream proxy.Upstream // this upstream is expected when testProxyCreator is called
+}
+
+func TestFallback(t *testing.T) {
+	// dummy Upstreams for servicing a specific rcode
+	dummyRefusedUpstream := &dummyUpstream{rcode: dns.RcodeRefused}
+	dummyServeFailureUpstream := &dummyUpstream{rcode: dns.RcodeServerFailure}
+	dummyNxDomainUpstream := &dummyUpstream{rcode: dns.RcodeNameError}
+	dummyUpstreams := []*dummyUpstream{
+		dummyRefusedUpstream,
+		dummyServeFailureUpstream,
+		dummyNxDomainUpstream,
+	}
+
+	testCases := []fallbackTestCase{
+		{
+			rcode:            dns.RcodeRefused,
+			expectedUpstream: dummyRefusedUpstream,
+		},
+		{
+			rcode:            dns.RcodeServerFailure,
+			expectedUpstream: dummyServeFailureUpstream,
+		},
+		{
+			rcode:            dns.RcodeNameError,
+			expectedUpstream: dummyNxDomainUpstream,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("rcode = %d", tc.rcode), func(t *testing.T) {
+			handler := newFallback(nil)
+			// create stub handler to return the test rcode
+			handler.Next = stubNextHandler(tc.rcode, nil)
+			// add dummyUpstreams to upstream map according to the rcode field
+			for _, u := range dummyUpstreams {
+				handler.rules[u.rcode] = u
+			}
+			proxyCreator := &testProxyCreator{t: t, expectedUpstream: tc.expectedUpstream}
+			handler.proxy = proxyCreator
+
+			ctx := context.TODO()
+			req := &dns.Msg{
+				Question: []dns.Question{{
+					Name:   "abc.com",
+					Qclass: dns.ClassINET,
+					Qtype:  dns.TypeA,
+				}},
+			}
+
+			rec := dnstest.NewRecorder(&test.ResponseWriter{})
+			_, _ = handler.ServeDNS(ctx, rec, req)
+
+			// Ensure that the proxyCreator is called once
+			if proxyCreator.called != 1 {
+				t.Fatalf("Expect proxy creator to be called once, but got '%d", proxyCreator.called)
+			}
+		})
+	}
+}
+
+func TestFallbackNotCalled(t *testing.T) {
```

Comment:
Why all this mocking and not just checking if the impl does the right thing if an upstream returns nxdomain (i.e. dnstest.Server should make this relatively painless)

Reply:
I must confess that I'm not familiar with dnstest. I just took a quick look and it seems, IMHO, that it looks more akin to integration test.

How about adding a couple of tests using dnstest to supplement?